hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -14,8 +32,296 @@ import HIE.Bios as BIOS
+  case filter (\p -> isCabalNewProject p || isStackProject p) projs of
+    (x:_) -> return $ Just x
+    []    -> case filter isCabalOldProject projs of
+      (x:_) -> return $ Just x
+      []    -> return Nothing
+
+    where
+      -- | Subdirectories of a given FilePath.
+      -- Directory closest to the FilePath `fp` is the head,
+      -- followed by one directory taken away.
+      subdirs :: [FilePath]
+      subdirs = reverse . map joinPath . tail . inits
+        $ splitDirectories (takeDirectory fp)
+
+      isStackProject (Ex ProjLocStackYaml {}) = True
+      isStackProject _ = False
+
+      isCabalNewProject (Ex ProjLocV2Dir {}) = True
+      isCabalNewProject (Ex ProjLocV2File {}) = True
+      isCabalNewProject _ = False
+
+      isCabalOldProject (Ex ProjLocV1Dir {}) = True
+      isCabalOldProject (Ex ProjLocV1CabalFile {}) = True
+      isCabalOldProject _ = False
+
+-- | Given a FilePath, find the Cradle the FilePath belongs to.
+--
+-- Finds the Cabal Package the FilePath is most likely a part of
+-- and creates a cradle whose root directory is the directory
+-- of the package the File belongs to.
+cabalHelperCradle :: FilePath -> IO Cradle
+cabalHelperCradle file = do
+  projM <- findCabalHelperEntryPoint file
+  case projM of
+    Nothing        -> do
+      errorm $ "Could not find a Project for file: " ++ file
+      error $ "Could not find a Project for file: " ++ file
+    Just (Ex proj) -> do
+      -- Find the root of the project based on project type.
+      let root = projectRootDir proj
+      -- Create a suffix for the cradle name.
+      -- Purpose is mainly for easier debugging.
+      let actionNameSuffix = projectSuffix proj
+      logm $ "Cabal-Helper dirs: " ++ show [root, file]
+      let dist_dir = getDefaultDistDir proj
+      env <- mkQueryEnv proj dist_dir
+      packages <- runQuery projectPackages env
+      -- Find the package the given file may belong to.
+      -- If it does not belong to any package, create a none-cradle.
+      -- We might want to find a cradle without actually loading anything.
+      -- Useful if we only want to determine a ghc version to use.
+      case packages `findPackageFor` file of
+        Nothing          -> do
+          debugm $ "Could not find a package for the file: " ++ file
+          debugm
+            "This is perfectly fine if we only want to determine the GHC version."
+          return
+            Cradle { cradleRootDir = root
+                   , cradleOptsProg =
+                       CradleAction { actionName = "Cabal-Helper-"
+                                        ++ actionNameSuffix
+                                        ++ "-None"
+                                    , runCradle = \_ -> return CradleNone
+                                    }
+                   }
+        Just realPackage -> do
+          debugm $ "Cabal-Helper cradle package: " ++ show realPackage
+          -- Field `pSourceDir` often has the form `<cwd>/./plugin`
```

Comment:
You should call `canonicalizePath ` from `System.Directory` to normalise a filepath I think.

Reply:
Doesnt normalise also work in this simple case?